tests/input_packet_test.cpp

```diff
@@ -0,0 +1,132 @@
+#include <iostream>
+#include <cassert>
+#include "../include/input_packet.h"
+
+void printInputPacket(const InputPacket &packet)
+{
+    std::cout << "Packet Type: ";
+    switch (packet.type)
+    {
+    case InputPacket::KEYBOARD:
+        std::cout << "KEYBOARD" << std::endl;
+        std::cout << "Key Code: " << packet.keyboard.keyCode << std::endl;
+        std::cout << "Pressed: " << (packet.keyboard.pressed ? "true" : "false") << std::endl;
+        break;
+    case InputPacket::MOUSE:
+        std::cout << "MOUSE" << std::endl;
+        std::cout << "X: " << packet.mouse.x << ", Y: " << packet.mouse.y << std::endl;
+        std::cout << "Button: " << packet.mouse.button << std::endl;
+        std::cout << "Pressed: " << (packet.mouse.pressed ? "true" : "false") << std::endl;
+        break;
+    case InputPacket::KEYBOARD_AND_MOUSE:
+        std::cout << "KEYBOARD_AND_MOUSE" << std::endl;
+        std::cout << "Key Code: " << packet.keyboard.keyCode << std::endl;
+        std::cout << "Key Pressed: " << (packet.keyboard.pressed ? "true" : "false") << std::endl;
+        std::cout << "X: " << packet.mouse.x << ", Y: " << packet.mouse.y << std::endl;
+        std::cout << "Button: " << packet.mouse.button << std::endl;
+        std::cout << "Button Pressed: " << (packet.mouse.pressed ? "true" : "false") << std::endl;
+        break;
+    default:
+        std::cout << "Unknown" << std::endl;
+        break;
+    }
+}
+
+bool compareInputPackets(const InputPacket &packet1, const InputPacket &packet2)
+{
+    if (packet1.type != packet2.type)
+        return false;
+    switch (packet1.type)
+    {
+    case InputPacket::KEYBOARD:
+        return packet1.keyboard.keyCode == packet2.keyboard.keyCode &&
+               packet1.keyboard.pressed == packet2.keyboard.pressed;
+    case InputPacket::MOUSE:
+        return packet1.mouse.x == packet2.mouse.x &&
+               packet1.mouse.y == packet2.mouse.y &&
+               packet1.mouse.button == packet2.mouse.button &&
+               packet1.mouse.pressed == packet2.mouse.pressed;
+    case InputPacket::KEYBOARD_AND_MOUSE:
+        return packet1.keyboard.keyCode == packet2.keyboard.keyCode &&
+               packet1.keyboard.pressed == packet2.keyboard.pressed &&
+               packet1.mouse.x == packet2.mouse.x &&
+               packet1.mouse.y == packet2.mouse.y &&
+               packet1.mouse.button == packet2.mouse.button &&
+               packet1.mouse.pressed == packet2.mouse.pressed;
+    default:
+        return false;
+    }
+}
+
+int main()
+{
+    // Test for KEYBOARD
+    {
+        InputPacket inputKeyboard;
+        inputKeyboard.type = InputPacket::KEYBOARD;
+        inputKeyboard.keyboard.keyCode = 65; // 'A'
+        inputKeyboard.keyboard.pressed = false;
+
+        char serializedData[sizeof(InputPacket)];
+        char *outData = serializedData;
```

Comment:
These point to the same addr, is that intentional? If so, why do we need two variables for it?

Reply:
outData is necessary as it modifies the buffer during serialization while keeping serializedData unchanged as a reference to the start of the buffer for correct deserialization.